src/main/java/net/mcreator/ui/minecraft/states/IPropertyData.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * MCreator (https://mcreator.net/)
+ * Copyright (C) 2012-2020, Pylo
+ * Copyright (C) 2020-2023, Pylo, opensource contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package net.mcreator.ui.minecraft.states;
+
+import net.mcreator.ui.MCreator;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.swing.*;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Instances of this class store information about certain property (its name, type,
+ * minimum and maximum values for number properties, list of allowed values for string properties and so on).
+ *
+ * @param <T> Type of values this property can take.
+ */
+public interface IPropertyData<T> {
+
+	/**
+	 * @return The name of this property.
+	 */
+	String getName();
+
+	/**
+	 * Provides the default value of type of this property. This is the "null" value of this type, which means
+	 * it may be outside value limits defined for a particular property.
+	 *
+	 * @return A default value of this property's type.
+	 */
+	@Nonnull T getDefaultValue();
+
+	/**
+	 * Checks if the provided value is different from {@code null} and its type matches type of this property
+	 * or its subtype. Falls back to the default property value if this condition is not met.
+	 *
+	 * @param value A value of this property's type.
+	 * @return Passed value, or the default one if {@code null}.
+	 */
+	@SuppressWarnings("unchecked") default T checkValue(Object value) {

```

Comment:
Why is this needed? When can type not match?

IF there is a case where this is needed, the root cause should be taken care of, not some hack-around be made

Reply:
If the value is null or of different type, it is replaced by the default type value here.

Reply:
This may be the case if the state is being created, the property component was previously disabled/excluded from the state, or if ME definition was corrupt and used values of inappropriate types.

Reply:
> ME definition was corrupt and used values of inappropriate types.

I can not imagine scenario where this would happen this way. Why don't we then check for every type when deserializing anything to make sure it won't corrupt?

Reply:
So `Objects.requireNonNullElse((T) value, getDefaultValue())` should be enough? Or this method should not exist at all?

Reply:
I did not ask for this change and there was no reason to make it, yet you made it, but did not make changes I asked you to.

The number type is still somehow called Num, where I asked for PropertyData.TypeNumber, mappings are still not used and the PR is marked as ready for review. 

Not sure what is going on in this PR, but this way we can not communicate and handle big PRs. Don't get me wrong, but by the time I had to go back and forth with this PR I would have made this PR myself

Reply:
Null value indicates the property was not used in a state.

Reply:
> You had check in place before that was fine.

Also this refactor now brought reviewed files to 5 as you had to refactor even more files

Reply:
TLDR: rever to how this was handled before without the use of this method

Reply:
✔️ Fixed